tools/job-cache/job-cache.cpp

```diff
@@ -0,0 +1,518 @@
+    bind_string(why, add_input_dir, 2, hash.c_str(), hash.size());
+    bind_integer(why, add_input_dir, 3, job_id);
+    single_step(why, add_input_dir);
+    finish_stmt(why, add_input_dir);
+  }
+
+  void insert_output_file(const std::string &path, const std::string &hash, int64_t job_id) {
+    const char *why = "Could not insert output file";
+    bind_string(why, add_output_file, 1, path.c_str(), path.size());
+    bind_string(why, add_output_file, 2, hash.c_str(), hash.size());
+    bind_integer(why, add_output_file, 3, job_id);
+    single_step(why, add_output_file);
+    finish_stmt(why, add_output_file);
+  }
+
+ public:
+  ~Cache() {
+    FINALIZE(add_job)
+    FINALIZE(add_input_file)
+    FINALIZE(add_input_dir)
+    FINALIZE(add_output_file)
+    FINALIZE(begin_txn_query)
+    FINALIZE(commit_txn_query)
+
+    if (sqlite3_close(db) != SQLITE_OK) {
+      std::cerr << "Could not close wake.db: " << sqlite3_errmsg(db) << std::endl;
+      return;
+    }
+  }
+
+  Cache() = delete;
+  Cache(const Cache &) = delete;
+  Cache(std::string _dir) : dir(_dir), rng(get_rng_seed()) {
+    mkdir_no_fail(_dir.c_str());
+    init_sql();
+    prepare_stmts();
+  }
+
+  void add(const JAST &job_result_json) {
+    // Create a unique name for the job dir (will rename later to correct name)
+    std::string tmp_job_dir = dir + "/tmp_" + rng.unique_name();
+    mkdir_no_fail(tmp_job_dir.c_str());
+
+    // Parse the json
+    AddJobRequest add_request(job_result_json);
+
+    // Copy the output files into the temp dir
+    for (const auto &output_file : add_request.outputs) {
+      std::string blob_path = tmp_job_dir + "/" + output_file.hash;
+      copy_or_reflink(output_file.source.c_str(), blob_path.c_str());
+    }
+
+    // Start a transaction so that a job is never without its files.
+    begin_txn();
+    uint64_t job_id = insert_job(add_request.cwd, add_request.command_line, add_request.envrionment,
+                                 add_request.stdin_str, add_request.bloom);
+
+    // Input Files
+    for (const auto &input_file : add_request.inputs) {
+      insert_input_file(input_file.path, input_file.hash, job_id);
+    }
+
+    // Input Dirs
+    for (const auto &input_dir : add_request.directories) {
+      insert_input_dir(input_dir.path, input_dir.hash, job_id);
+    }
+
+    // Output Files
+    for (const auto &output_file : add_request.outputs) {
+      insert_input_dir(output_file.path, output_file.hash, job_id);
+    }
+
+    // TODO: Add clean up on init to `rm -rf` the tmp_ jobs
+    //       in case we fail before we rename
+    uint8_t job_group = job_id & 0xFF;
+    std::string job_group_dir = dir + "/" + to_hex<uint8_t>(&job_group);
+    mkdir_no_fail(job_group_dir.c_str());
+    std::string job_dir = job_group_dir + "/" + std::to_string(job_id);
+    rename_no_fail(tmp_job_dir.c_str(), job_dir.c_str());
+
+    // TODO: Add WAL to handle case where we fail here. We
+    //       can just append the job_id to a file.
```

Comment:
Actully I spoke with Wes about this and I don't need to do this, but I do need to have the transaction *end* before the atomic rename and I need to refanagal the renames a bit. I'll do that in this change.